src/rust/engine/engine_cffi/src/lib.rs

```diff
@@ -899,7 +906,29 @@ pub extern "C" fn materialize_directories(
           })
           .collect::<Vec<_>>(),
       )
-      .map(|_| ()),
+      .map(move |metadata: Vec<store::DirectoryMaterializeMetadata>| {
```

Comment:
I'm a fan of type annotations in about any position, even when not required. That said, there appear to be several here that aren't required and it runs against the grain of the existing style. Consider removing them now that you've debugged the compile if that's what they were added for.

Reply:
I left the annotation on the `.map` since I think it's useful for reminding anyone who looks at this code in the future what rust type is involved, makes for easier grepping. Removed the one you pointed out.